src/rgw/rgw_datalog.h

```diff
@@ -125,6 +125,40 @@ struct RGWDataChangesLogMarker {
   RGWDataChangesLogMarker() = default;
 };
 
+struct rgw_data_notify_entry {
+  std::string key;
+  uint64_t gen = 0;
+
+  rgw_data_notify_entry(const std::string key, uint64_t gen): key(key), gen(gen) {}
+
+  rgw_data_notify_entry(const std::string& strKey = "")
+  : key(strKey) {}
+
+  void encode(ceph::buffer::list& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(key, bl);
+    encode(gen, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(ceph::buffer::list::const_iterator& bl) {
+     DECODE_START(1, bl);
+     decode(key, bl);
+     decode(gen, bl);
+     DECODE_FINISH(bl);
+  }
+
+  void dump(ceph::Formatter* f) const;
+  void decode_json(JSONObj* obj);
+
+  rgw_data_notify_entry& operator=(const rgw_data_notify_entry&) = default;
+
+  bool operator<(const rgw_data_notify_entry& d) const {
+    return (key < d.key && gen < d.gen);
```

Comment:
when sorting something with two fields, you usually pick one for the primary sort order, and only consider the second field for entries with the same primary

here's an example from rgw_bucket_shard in rgw_basic_types.h:
```c++
  bool operator<(const rgw_bucket_shard& b) const {
    if (bucket < b.bucket) {
      return true;
    }
    if (b.bucket < bucket) {
      return false;
    }
    return shard_id < b.shard_id;
  }
```
i'd choose `key` as the primary here so it'd sort like this: `a:0, a:1, b:3, c:0, d:0, d:1`